### PR TITLE
Measure/ValueSet object updates

### PR DESCRIPTION
**Pull request by okeefm**
This Pull Request fixes a number of issues on the `Valueset`, `Concept`, `Measure`, and `MeasurePackage` objects:

* Objects that should not be in the `QDM` namespace (because they're not QDM datatypes) have been moved into the `Tacoma` module in Ruby
* Changed `mString`, `mBoolean`, etc. to `String`, `Boolean` etc. where necessary on the JS side, so no more `mString`-like variable names should exist.
* Commented various `Measure` ID fields
* Removed `measure_id` as it seemed duplicative of other ID fields. Bonnie reviewer should review this and ensure it's true. If it is duplicative, we can create a method/ called `measure_id` that returns the other field.
* Changed `Valueset` handling in `Measure`, and field names in `Valueset`/`Concept` to comply with @hossenlopp's documentation
* Removed a number of Cypress/Bonnie specific fields, in favor of adding those in the specific projects where needed.
* Made a number of changes requested by @hossenlopp in his Measure object review. NOTE: There are still a couple of `Hash`/`Mixed` object refactors in progress that will be added here.

Pull requests into cqm-models require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable ("N/A") and :white_check_mark:.

**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] Internal ticket for this PR: N/A
- [x] Internal ticket links to this PR N/A
- [x] Code diff has been done and been reviewed
- [x] Tests are included and test edge cases N/A
- [x] Tests have been run locally and pass
- [x] If there were any JavaScript changes, this PR has updated the `dist` directory using `npm run dist`
- [x] If applicable, the library version number in `package.json` and `cqm-models.gemspec` has been updated

**Bonnie Reviewer:**

Name: @hossenlopp 
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

**Cypress Reviewer:**

Name: @dczulada
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

### CODE REVIEW

File: app/assets/javascripts/Measure.js [hunk 1]
```diff
@@ -14,27 +14,23 @@ const [Number, String, Boolean, Mixed, ObjectId, Date] = [
...
     cms_id: String,
     title: String,
     description: String,
+    // EP/EH, or "unknown"
     type: String,
     category: { type: String, default: 'Uncategorized' },
 
     // Measure type variables
     episode_of_care: Boolean,
     continuous_constiable: Boolean,
```
Comment (hossenlopp) [inline app/assets/javascripts/Measure.js]: Huh? `continuous_variable`?

File: app/models/qdm/tacoma/measure.rb [hunk 2]
```diff
@@ -64,34 +58,29 @@ class Measure
...
+    # about getting all the measures from a ValueSet object,
+    # hence the 'inverse_of: nil')
+    has_and_belongs_to_many :value_sets, inverse_of: nil
 
     scope :by_measure_id, ->(id) { where('measure_id' => id) }
     scope :by_type, ->(type) { where('type' => type) }
     scope :by_user, ->(user) { where user_id: user.id }
 
     index 'user_id' => 1
+    index 'hqmf_id' => 1
```
Comment (hossenlopp) [inline app/models/qdm/tacoma/measure.rb]: `hqmf_id` wouldn't be a useful index.

A compound index on `user_id` and `hqmf_set_id` would be useful since it's the most common query in bonnie.
`index 'user_id' => 1, 'hqmf_set_id' => 1`

File: app/models/qdm/tacoma/valueset.rb [hunk 3]
```diff
@@ -1,19 +1,17 @@
...
     field :version, type: String
-    field :user_id, type: String # Eventually we need to delete this from bundles when exporting
-    field :categories, type: Hash
 
     index(oid: 1)
+
     embeds_many :concepts
     index 'concepts.code' => 1
-    index 'concepts.code_system' => 1
+    index 'concepts.code_system_oid' => 1
```
Comment (hossenlopp) [inline app/models/qdm/tacoma/valueset.rb]: In my opinion, indexes on the concepts aren't useful at all. We are never looking for concepts in particular.

File: app/models/qdm/tacoma/valueset.rb [hunk 4]
```diff
@@ -1,19 +1,17 @@
...
   # ValueSet represents a collection of Concepts, used by the Measures to specify a set of codes for a particular topic
   class ValueSet
     include Mongoid::Document
     field :oid, type: String
     field :display_name, type: String
     field :version, type: String
-    field :user_id, type: String # Eventually we need to delete this from bundles when exporting
-    field :categories, type: Hash
 
     index(oid: 1)
```
Comment (hossenlopp) [inline app/models/qdm/tacoma/valueset.rb]: Doesn't match hash style of other index calls.

File: app/assets/javascripts/Measure.js (hunk 1 above)
Comment (okeefm) [reply app/assets/javascripts/Measure.js]: That was fixed...not sure how it crept back in.

File: app/models/qdm/tacoma/valueset.rb (hunk 4 above)
Comment (okeefm) [reply app/models/qdm/tacoma/valueset.rb]: Fixed.

File: app/assets/javascripts/IndividualResult.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    STRAT: Number,
+    IPP: Number,
+    DENOM: Number,
+    NUMER: Number,
+    NUMEX: Number,
+    DENEX: Number,
+    DENEXCEP: Number,
+    MSRPOPL: Number,
+    OBSERV: Number,
+    MSRPOPLEX: Number,
```
Comment (edeyoung) [inline app/assets/javascripts/IndividualResult.js]: Didn't we decide that there would be two versions of this? one for pop based and one for episode based? pop based would be boolean, episode based would be *the array of episodes*?

Also, doesn't OBSERV need to be an array of numbers?

File: app/assets/javascripts/Measure.js [hunk 6]
```diff
@@ -14,33 +14,29 @@ const [mNumber, mString, mBoolean, Mixed, ObjectId, mDate] = [
...
+    // A version-neutral UUID for the measure
+    hqmf_set_id: String,
+    // A Semantic Version-compliant string (e.g. "2.3.4") for the measure
+    hqmf_version_number: String,
+    // A CMS-style string (e.g. "CMS2v4") for the measure
+    cms_id: String,
+    title: String,
+    description: String,
+    // EP/EH, or "unknown"
+    type: String,
```
Comment (edeyoung) [inline app/assets/javascripts/Measure.js]: comment should fully list the options.

"EH", "EP", "unknown" (or whatever these actually are)

Should this be an enum?

File: app/assets/javascripts/Measure.js [hunk 7]
```diff
@@ -14,33 +14,29 @@ const [mNumber, mString, mBoolean, Mixed, ObjectId, mDate] = [
...
+    category: { type: String, default: 'Uncategorized' },
```
Comment (edeyoung) [inline app/assets/javascripts/Measure.js]: add comment for what category means.

File: app/assets/javascripts/Measure.js [hunk 8]
```diff
@@ -53,14 +49,12 @@ const MeasureSchema = mongoose.Schema(
...
-
     // Relations to other model classes
-    user: { type: ObjectId, ref: 'User', index: true }, // Bonnie-specific
-    bundle: { type: ObjectId, ref: 'Bundle' }, // Cypress-specific, but used in Bonnie-bundler as well
-    package: { type: ObjectId, ref: 'MeasurePackage' }, // Bonnie specific
+    bundle: { type: ObjectId, ref: 'Bundle' }, // Cypress-specific, until we migrate the Bundle into cqm-models
+    package: { type: ObjectId, ref: 'MeasurePackage' }, // Bonnie-specific
+    patients: [{ type: ObjectId, ref: 'Patient', index: true }],
+
+    value_sets: [{ type: ObjectId, ref: 'ValueSet' }],
```
Comment (edeyoung) [inline app/assets/javascripts/Measure.js]: does this align w the new model for handling value sets?

File: app/assets/javascripts/Measure.js [hunk 9]
```diff
@@ -14,33 +14,29 @@ const [mNumber, mString, mBoolean, Mixed, ObjectId, mDate] = [
...
-    published: mBoolean,
-    publish_date: mDate,
-    version: mNumber,
+    episode_of_care: Boolean,
+    continuous_variable: Boolean,
 
     // ELM/CQL Measure-logic related data
     elm_annotations: Mixed,
-    cql: [mString],
+    cql: [String],
```
Comment (edeyoung) [inline app/assets/javascripts/Measure.js]: see comment here: https://github.com/projecttacoma/cqm-models/pull/10/files#diff-c25f2a7e44d249869432e81248758e6eR41

Is this going to be updated?

File: app/assets/javascripts/Measure.js (hunk 6 above)
Comment (okeefm) [reply app/assets/javascripts/Measure.js]: These are the only options I saw in Bonnie, other than `null`, when I ran a `db.getCollection('cql_measures').distinct('type')`. I don't know that the type is (or should be) restricted to only those options, but I'd be willing to change it to be an enum if others want me to.

File: app/assets/javascripts/Measure.js (hunk 7 above)
Comment (okeefm) [reply app/assets/javascripts/Measure.js]: Will do.

File: app/assets/javascripts/Measure.js [hunk 10]
```diff
@@ -14,33 +14,29 @@ const [mNumber, mString, mBoolean, Mixed, ObjectId, mDate] = [
...
-    episode_of_care: mBoolean,
-    continuous_constiable: mBoolean,
-    episode_ids: [],
-
-    // Publishing data (used by Bonnie)
-    published: mBoolean,
-    publish_date: mDate,
-    version: mNumber,
+    episode_of_care: Boolean,
+    continuous_variable: Boolean,
```
[the remark this answers is not in the document]
Comment (okeefm) [reply app/assets/javascripts/Measure.js]: What other options will be required here? I'm not sure I'm familiar enough with the new "odd" measures to know what should go here. Should I talk to @hossenlopp or @jbradl11 ?

File: app/assets/javascripts/Measure.js (hunk 8 above)
Comment (okeefm) [reply app/assets/javascripts/Measure.js]: This does, yes. @hossenlopp's document has Measure objects having many `ValueSet` objects, and this is how you do that in Mongoose (a non-embedded many relation).

File: app/assets/javascripts/Measure.js (hunk 6 above)
Comment (edeyoung) [reply app/assets/javascripts/Measure.js]: Were the options `["EP", "EH", "unknown"]` or `["EP/EH", "unknown"]`?
It is not clear to me from the comment which it is - the comment is requesting that this be clarified.
If it's the latter, please put "EP/EH" in quotes.
It should be explicit to the user what the options are.

Regarding whether it should be in an enum or not, that can be something that we take a look at later. I think it should be - but i agree we should confirm what the total possible options are before we limit things.

File: app/assets/javascripts/Measure.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (edeyoung) [reply app/assets/javascripts/Measure.js]: for now, don't add any "odd" measures in.
measure types are:

* PROPORTION
* RATIO
* CONTINOUS_VARIABLE

Calculation methods are:

* PATIENT
* EPISODE_OF_CARE

File: app/assets/javascripts/Measure.js (hunk 6 above)
Comment (okeefm) [reply app/assets/javascripts/Measure.js]: Ah, I'm sorry I didn't get that the first time. I'll make that more clear from the comment.

Comment (edeyoung) [reply app/assets/javascripts/Measure.js]: this doesn't look like it was clarified

File: app/assets/javascripts/Measure.js [hunk 11]
```diff
@@ -2,46 +2,49 @@ const mongoose = require('mongoose');
...
   mongoose.Schema.Types.Mixed,
   mongoose.Schema.Types.ObjectId,
   mongoose.Schema.Types.Date,
 ];
 
 const MeasureSchema = mongoose.Schema(
   {
-    // ID/other measure information
-    id: mString,
-    measure_id: mString,
```
Comment (dczulada) [inline app/assets/javascripts/Measure.js]: measure_id is used in "as_hqmf_model"

Comment (okeefm) [reply app/assets/javascripts/Measure.js]: Good catch. It seems like `measure_id` is always equivalent to `hqmf_id` when set; are you ok with us using `hqmf_id` in place of `measure_id` in the `as_hqmf_model` method?

File: app/assets/javascripts/Measure.js [hunk 12]
```diff
@@ -14,33 +14,29 @@ const [mNumber, mString, mBoolean, Mixed, ObjectId, mDate] = [
...
+    continuous_variable: Boolean,
 
     // ELM/CQL Measure-logic related data
     elm_annotations: Mixed,
-    cql: [mString],
+    cql: [String],
     elm: [Mixed],
-    main_cql_library: mString,
+    main_cql_library: String,
     cql_statement_dependencies: Mixed,
```
[the remark this answers is not in the document]
Comment (hossenlopp) [reply app/assets/javascripts/Measure.js]: @okeefm and I can plan on defining this better when we work on the population set structure. We could use the same type defined for library/statement references that we plan on using for the statement population references.
